Userland/Libraries/LibCore/MachPort.cpp

```diff
@@ -0,0 +1,177 @@
+{
+    switch (right) {
+    case MachPort::MessageRight::MoveReceive:
+        return MACH_MSG_TYPE_MOVE_RECEIVE;
+    case MachPort::MessageRight::MoveSend:
+        return MACH_MSG_TYPE_MOVE_SEND;
+    case MachPort::MessageRight::MoveSendOnce:
+        return MACH_MSG_TYPE_MOVE_SEND_ONCE;
+    case MachPort::MessageRight::CopySend:
+        return MACH_MSG_TYPE_COPY_SEND;
+    case MachPort::MessageRight::MakeSend:
+        return MACH_MSG_TYPE_MAKE_SEND;
+    case MachPort::MessageRight::MakeSendOnce:
+        return MACH_MSG_TYPE_MAKE_SEND_ONCE;
+    case MachPort::MessageRight::CopyReceive:
+        return MACH_MSG_TYPE_COPY_RECEIVE;
+    case MachPort::MessageRight::DisposeReceive:
+        return MACH_MSG_TYPE_DISPOSE_RECEIVE;
+    case MachPort::MessageRight::DisposeSend:
+        return MACH_MSG_TYPE_DISPOSE_SEND;
+    case MachPort::MessageRight::DisposeSendOnce:
+        return MACH_MSG_TYPE_DISPOSE_SEND_ONCE;
+    }
+    VERIFY_NOT_REACHED();
+}
+
+static Error mach_error_to_error(kern_return_t error)
+{
+    char const* err_string = mach_error_string(error);
+    StringView const err_view(err_string, strlen(err_string));
+    return Error::from_string_view(err_view);
+}
+
+static Error bootstrap_error_to_error(kern_return_t error)
+{
+    char const* err_string = bootstrap_strerror(error);
+    StringView const err_view(err_string, strlen(err_string));
+    return Error::from_string_view(err_view);
+}
+
+MachPort::MachPort(PortRight right, mach_port_t port)
+    : m_right(right)
+    , m_port(port)
+{
+}
+
+MachPort::~MachPort()
+{
+    unref_port();
+}
+
+MachPort::MachPort(MachPort&& other)
+    : m_port(other.release())
+{
+}
+
+MachPort& MachPort::operator=(MachPort&& other)
+{
+    if (this != &other) {
+        unref_port();
+        m_port = other.release();
+    }
+    return *this;
+}
+
+void MachPort::unref_port()
+{
+    if (!MACH_PORT_VALID(m_port))
+        return;
+
+    switch (m_right) {
+    case PortRight::Send:
+    case PortRight::SendOnce:
+    case PortRight::DeadName:
+        mach_port_deallocate(mach_task_self(), m_port);
+        break;
+    case PortRight::Receive:
+    case PortRight::PortSet:
+        mach_port_mod_refs(mach_task_self(), m_port, to_native_right(m_right), -1);
+        break;
+    }
+}
+
+ErrorOr<MachPort> MachPort::create_with_right(PortRight right)
+{
+    mach_port_t port = MACH_PORT_NULL;
+    auto const ret = mach_port_allocate(mach_task_self(), to_native_right(right), &port);
+    if (ret != KERN_SUCCESS) {
+        dbgln("Unable to allocate port with right: {}", to_native_right(right));
+        return mach_error_to_error(ret);
+    }
+    return MachPort(right, port);
+}
+
+MachPort MachPort::create_from_received_port(mach_port_t port, PortRight right)
+{
+    return MachPort(right, port);
+}
+
+mach_port_t MachPort::release()
+{
+    return exchange(m_port, MACH_PORT_NULL);
+}
+
+ErrorOr<void> MachPort::insert_message_right(MessageRight right)
```

Comment:
Er, doesn't this take the mach_port_t I already had and attach more stuff to it? I was assuming this worked like a fcntl or setsockopt

Reply:
What `mach_port_insert_right()` really does is it inserts a port right into an IPC namespace of a task. You'll notice it takes `m_port` twice: the first is the name you'd like to insert the right into the target task under, and the second is the name in the caller's namespace. In your case, the target task and the caller task are the same, `mach_task_self()`, so the names have to match.

So in a way, yes, it's just attaching more stuff to an existing port. But something still has to track ownership of the newly inserted right; `this` won't, since `m_right` still specifies e.g. `Receive`. When you destroy `this`, it will destroy the receive right, but the send right will leak and live on (as a dead name, once the receive right gets destroyed).

Reply:
> I was assuming this worked like a fcntl or setsockopt

You should rather think of it as being like `dup()` or something. You get a new right that has to be released independently, it just happens to share the same name.

Reply:
This means I'll have to translate between the right I'm inserting and the resulting port right, right? should be straightforward either way.

Reply:
I don't think I understand what you mean by "translate" here.